### ISSUE TITLE
GPU utilization, latency during inference and model accuracy

### ISSUE THREAD

**Issue by ishanic** [facebookresearch/meshrcnn#16]
Hello,
I trained a model on my 2-GPU machine, basically to reproduce the results of the meshrcnn_R50 that's shared here, without modifying anything. 
>> python tools/train_net.py --num-gpus 2 --config-file configs/pix3d/meshrcnn_R50_FPN.yaml

I ran the evaluation both with the shared meshrcnn_R50.pth and the model I trained, and it seems that the accuracy of the model is 10% worse using my model. What's stranger is that the GPU utilization and the time for inference is 4x worse. Below, I have shared the stats for inference on 8 images from Pix3D.

>> time python tools/train_net.py --num-gpus 2 --config-file configs/pix3d/meshrcnn_R50_FPN.yaml --eval-only MODEL.WEIGHTS output/meshrcnn_R50.pth

real	0m15.656s
user	0m35.867s
sys	0m7.237s
[02/28 18:22:14 meshrcnn.evaluation.pix3d_evaluation]: Box AP 0.92593
[02/28 18:22:14 meshrcnn.evaluation.pix3d_evaluation]: Mask AP 0.92593
[02/28 18:22:14 meshrcnn.evaluation.pix3d_evaluation]: Mesh AP 0.44444
GPU util: 2.8-2.9GB per machine, as per nvidia-smi

>> time python tools/train_net.py --num-gpus 2 --config-file configs/pix3d/meshrcnn_R50_FPN.yaml --eval-only MODEL.WEIGHTS output/model_final.pth

real	0m35.237s
user	1m5.214s
sys	0m9.366s
[02/28 18:20:49 meshrcnn.evaluation.pix3d_evaluation]: Box AP 0.86772
[02/28 18:20:49 meshrcnn.evaluation.pix3d_evaluation]: Mask AP 0.75661
[02/28 18:20:49 meshrcnn.evaluation.pix3d_evaluation]: Mesh AP 0.40432
GPU util: 8GB per machine, as per nvidia-smi

Please let me know if you have any ideas of why this might happen. I have also shared the detectron2 collect_env output below

---
sys.platform              linux
Python                    3.7.6 (default, Jan  8 2020, 19:59:22) [GCC 7.3.0]
numpy                     1.17.4
detectron2                0.1 @/home/USER/detectron2/detectron2
detectron2 compiler       GCC 5.5
detectron2 CUDA compiler  10.1
detectron2 arch flags     sm_50, sm_52, sm_60, sm_61, sm_70, sm_75
DETECTRON2_ENV_MODULE     <not set>
PyTorch                   1.4.0 @/home/USER/.conda/envs/meshrcnn1/lib/python3.7/site-packages/torch
PyTorch debug build       False
CUDA available            True
GPU 0                     TITAN X (Pascal)
GPU 1                     GeForce GTX TITAN X
CUDA_HOME                 /usr/local/cuda
NVCC                      Cuda compilation tools, release 10.1, V10.1.243
Pillow                    7.0.0
torchvision               0.5.0 @/home/USER/.conda/envs/meshrcnn1/lib/python3.7/site-packages/torchvision
torchvision arch flags    sm_35, sm_50, sm_60, sm_70, sm_75
cv2                       4.2.0
------------------------  -----------------------------------------------------------------------------------
PyTorch built with:
  - GCC 7.3
  - Intel(R) Math Kernel Library Version 2020.0.0 Product Build 20191122 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v0.21.1 (Git Hash 7d2fd500bc78936d1d648ca713b901012f470dbc)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - NNPACK is enabled
  - CUDA Runtime 10.1
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_61,code=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_37,code=compute_37
  - CuDNN 7.6.3
  - Magma 2.5.1
  - Build settings: BLAS=MKL, BUILD_NAMEDTENSOR=OFF, BUILD_TYPE=Release, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -fopenmp -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -O2 -fPIC -Wno-narrowing -Wall -Wextra -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Wno-stringop-overflow, DISABLE_NUMA=1, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, USE_CUDA=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, USE_STATIC_DISPATCH=OFF,

**Comment by gkioxari** [comment]
Hi @ishanic. You say 

> I trained a model on my 2-GPU machine

Unfortunately, you are deviating from the training recipe. Mesh R-CNN models, similar to Mask R-CNN, are trained using 8-GPUs and the training recipe is designed for this setup. Once you deviate from that, there is no guarantee performance will be kept. In your case, training with 2 GPUs means you are training with batches which are 4 times smaller. So it is completely expected that you can't reach the best performance. Hyperparameter tuning is important in deep learning! Read [this paper](https://arxiv.org/abs/1706.02677) for more. So here is some ways that *might* help bring back points in performance when training with 2 GPUs. (a) Lower the learning rate (try linearly first), (b) Stretch the training schedule by 4x, this means train for 4x longer iterations and also adjust the points at which learning drops accordingly (4x) - this will result in training for the same number of epochs. 

Now regarding the fact that the test times diverge between your model and mine I truly don't know why that is the case unless you have changed the model.

**Comment by ishanic** [comment]
Thank you for the detailed response. I had earlier fixed the learning rate alone, but now I modified several other parameters in the solver such as the # iterations, step size, warm up iterations etc. according to the linear scaling rule. These changes have reduced the gap. My model's accuracy matches in terms of box and mask APs, but still lags behind in the Mesh AP by 6%. Further tweaking the params should close the gap. 
Interestingly, the GPU memory usage and latency also reduced with a better trained model, probably due to sparsification of the weights?
I will close this issue now. Thanks!

**Comment by gkioxari** [comment]
Great! Unfortunately, there is no science or conclusive evidence on how to set hyperparameters. Mesh prediction is much harder than box and mask prediction so I can see how the gap is still positive for this task. I think tweaking the schedule could close the gap further. Let us know what you find as I think many folks would find it useful!

**Comment by albertotono** [comment]
Hi @ishanic, I hope your exploration went well. 

I was wondering if you could share with us your **config** for Pix3D with 2GPUs, did you also try with only 1? 

best Regards
